### ISSUE TITLE
Support for overrided PropertyNamingPolicy in JsonSerializerOptions

### ISSUE THREAD

**Issue by NSentinel** [dahomey-technologies/Dahomey.Json#47]
Library version `1.7.1` (and so on, including the latest `1.8.0`) with resolving [#45](https://github.com/dahomey-technologies/Dahomey.Json/issues/45) also breaks support for the non-standard naming policy.

We override `PropertyNamingPolicy` to support snake case naming. I think the library can't properly map names with underscores to the original property. 

.NET Fiddle: https://dotnetfiddle.net/qkJsUs

Sample:

```csharp
using System;
using System.Globalization;
using System.Text;
using System.Text.Json;
using Dahomey.Json;

public struct FirmwareFileInfo {
	public FirmwareFileInfo(byte fileType)
	{
		FileType = fileType;
	}
	
	public byte FileType { get; set; }
}

public class Program
{
	public static void Main()
	{
		var s1 = new FirmwareFileInfo();
		
		var options = new JsonSerializerOptions{
			PropertyNamingPolicy = JsonSnakeCaseNamingPolicy.Default
		};
		options.SetupExtensions();
		
		var json = JsonSerializer.Serialize(s1, options);
		Console.WriteLine("Serialized: {0}", json);
			
		var s2 = JsonSerializer.Deserialize<FirmwareFileInfo>(json, options);
		Console.WriteLine("Deserialized: {0}", s2.FileType);
	}
}

   public sealed class JsonSnakeCaseNamingPolicy : JsonNamingPolicy
    {
        public static readonly JsonSnakeCaseNamingPolicy Default = new JsonSnakeCaseNamingPolicy();

        public override string ConvertName(string name)
        {
            if (string.IsNullOrEmpty(name))
                return name;

            StringBuilder builder = new StringBuilder(name.Length + Math.Max(2, name.Length / 5));
            UnicodeCategory? previousCategory = null;

            for (int currentIndex = 0; currentIndex < name.Length; currentIndex++)
            {
                char currentChar = name[currentIndex];
                if (currentChar == '_')
                {
                    builder.Append('_');
                    previousCategory = null;
                    continue;
                }

                UnicodeCategory currentCategory = char.GetUnicodeCategory(currentChar);

                switch (currentCategory)
                {
                    case UnicodeCategory.UppercaseLetter:
                    case UnicodeCategory.TitlecaseLetter:
                        if (previousCategory == UnicodeCategory.SpaceSeparator ||
                            previousCategory == UnicodeCategory.LowercaseLetter ||
                            previousCategory != UnicodeCategory.DecimalDigitNumber &&
                            currentIndex > 0 &&
                            currentIndex + 1 < name.Length &&
                            char.IsLower(name[currentIndex + 1]))
                        {
                            builder.Append('_');
                        }

                        currentChar = char.ToLower(currentChar);
                        break;

                    case UnicodeCategory.LowercaseLetter:
                    case UnicodeCategory.DecimalDigitNumber:
                        if (previousCategory == UnicodeCategory.SpaceSeparator)
                        {
                            builder.Append('_');
                        }

                        break;

                    case UnicodeCategory.Surrogate:
                        break;

                    default:
                        if (previousCategory != null)
                        {
                            previousCategory = UnicodeCategory.SpaceSeparator;
                        }

                        continue;
                }

                builder.Append(currentChar);
                previousCategory = currentCategory;
            }

            return builder.ToString();
        }
	}
```

Execution result:

```
Serialized: {"file_type":0}

Unhandled exception. System.NotSupportedException: Specified method is not supported.
   at Dahomey.Json.Serialization.Converters.StructMemberConverter`2.Set(Object obj, Object value, JsonSerializerOptions options)
   at Dahomey.Json.Serialization.Converters.ObjectConverter`1.Read(Utf8JsonReader& reader, T& obj, JsonSerializerOptions options)
   at Dahomey.Json.Serialization.Converters.ObjectConverter`1.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options)
   at System.Text.Json.JsonPropertyInfoNotNullable`4.OnRead(ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.JsonPropertyInfo.Read(JsonTokenType tokenType, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
   at System.Text.Json.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.Deserialize(String json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
   at Program.Main()
```

**Comment by CXuesong** [comment]
Actually camelCase does not work. This snippet is a variation from #45 in the way of initializing `JsonSerializerOptions`.

https://dotnetfiddle.net/YPMM3U

```cs
public readonly struct MyStruct {
	[JsonConstructor("Value")]
	public MyStruct(int value) {
		Value = value;
	}
	
	public int Value { get; }
}

public class Program
{
	public static void Main()
	{
		var s1 = new MyStruct(100);
		var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true, PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
		options.SetupExtensions();
		var json = JsonSerializer.Serialize(s1, options);
		Console.WriteLine(json);
		var s2 = JsonSerializer.Deserialize<MyStruct>(json, options);
		Console.WriteLine("Deserialized value: {0}", s2.Value);
	}
}
```

Output is (expect 100 as deserialized value):
```
{"value":100}
Deserialized value: 0
```

**Comment by NSentinel** [comment]
Yeah, I have to check non-default initializer before thinking it works as expected and not rely on not throwing an exception as an indication.

**Comment by CXuesong** [comment]
I think there is another question: what does the ParamArray passed in `JsonConstructorAttribute` stands for. If I change the constructor in my sample into this
```cs
	[JsonConstructor("value")]
	public MyStruct(int value) {
```
then everything works.

Thus for now I assume the answer to the question should be "the serialized JSON property names". In that case, your `FirmwareFileInfo.ctor` should (but actually not, currently) imply the following, if you are using camel_case:
```cs
[JsonConstructor("file_type")]
public FirmwareFileInfo(byte fileType)
```
@NSentinel Actually if you apply the snippet above, your case should work.

**Comment by CXuesong** [comment]
Then the issue would be: let `CreatorMapping` aware of the `PropertyNamingPolicy` you are currently using somehow, if you do not explicitly apply `JsonConstructorAttribute`.

**Comment by NSentinel** [comment]
@CXuesong Thanks for the hint. I've tried before opening the issue but can't make it work in our real case. I oversimplified actual code when posting here to demonstrate the issue. 

We are using `Utf8Json` and trying to switch to "standard one" `System.Text.Json` in development. This library closes almost all incompatibilities for us. We can downgrade to `1.7.0` since it already works stable for us or if I can find time to dig issue maybe I'll fix it later and post here.  @mcatanzariti usually pretty fast in fixing, so probably he will be first :)
